decryptor/mysql/data_encoder.go

```diff
@@ -0,0 +1,388 @@
+package mysql
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"encoding/binary"
+	"errors"
+	"strconv"
+
+	"github.com/cossacklabs/acra/decryptor/base"
+	"github.com/cossacklabs/acra/encryptor"
+	"github.com/cossacklabs/acra/encryptor/config"
+	"github.com/cossacklabs/acra/encryptor/config/common"
+	"github.com/cossacklabs/acra/logging"
+	"github.com/sirupsen/logrus"
+)
+
+// ErrConvertToDataType error that indicates if data type conversion was failed
+var ErrConvertToDataType = errors.New("error on converting to data type")
+
+// BaseMySQLDataProcessor implements processor and encode/decode binary intX values to text format which acceptable by Tokenizer
+type BaseMySQLDataProcessor struct{}
+
+// DataEncoderProcessor implements processor and encode/decode binary intX values to text format which acceptable by Tokenizer
+type DataEncoderProcessor struct {
+	BaseMySQLDataProcessor
+}
+
+// NewDataEncoderProcessor return new data encoder from/to binary format for tokenization
+func NewDataEncoderProcessor() *DataEncoderProcessor {
+	return &DataEncoderProcessor{
+		BaseMySQLDataProcessor{},
+	}
+}
+
+// ID return name of processor
+func (p *DataEncoderProcessor) ID() string {
+	return "DataEncoderProcessor"
+}
+
+// DataDecoderProcessor implements processor and encode/decode binary intX values to text format which acceptable by Tokenizer
+type DataDecoderProcessor struct {
+	BaseMySQLDataProcessor
+}
+
+// NewDataDecoderProcessor return new data encoder from/to binary format for tokenization
+func NewDataDecoderProcessor() *DataDecoderProcessor {
+	return &DataDecoderProcessor{
+		BaseMySQLDataProcessor{},
+	}
+}
+
+// ID return name of processor
+func (p *DataDecoderProcessor) ID() string {
+	return "DataDecoderProcessor"
+}
+
+// here we process encryption/tokenization results before send it to a client
+// acra decrypts or de-tokenize SQL literals, so we should convert string SQL literals to binary format
+// if client expects int, then parse INT literals and convert to binary 4/8 byte format
+// if expects bytes, then pass as is
+// if expects string, then leave as is if it is valid string or encode to hex
+// if it is encrypted data then we return default values or as is if applicable (binary data)
+func (p *BaseMySQLDataProcessor) encodeBinary(ctx context.Context, data []byte, setting config.ColumnEncryptionSetting, columnInfo base.ColumnInfo, logger *logrus.Entry) (context.Context, []byte, error) {
+	logger = logger.WithField("column", setting.ColumnName()).WithField("decrypted", base.IsDecryptedFromContext(ctx))
+	logger.Debugln("Encode binary")
+
+	if len(data) == 0 {
+		// we still need to encode result data as it might be null field in db
+		return ctx, PutLengthEncodedString(data), nil
+	}
+
+	var columnType = columnInfo.DataBinaryType()
```

Comment:
let's assign this variable a bit below close to the usage block? because read this row and tried to find where it used

Reply:
Agree, makes sense.